[ios/cmakebuildsys] Add features to cmakebuildsys + jenkins ios xcodebuild

fuzzard opened the pull request:
## Description
Adds generator option (GEN) for cmakebuildsys.
Default ios generator to Xcode
Add option for user to provide build dir for cmakebuildsys
Updated build docs to use cmakebuildsys new options
binary-addon added to xcode project generation, docs updated how to selectively build addons

Updates Jenkins buildsteps to use xcodebuild when building ios

## Motivation and Context
Simplify build instructions to not require changing paths every time a new sdk/ios version comes out
Provides some customisation options for cmake project generation

## How Has This Been Tested?
Complete build using instructions in doc exactly

## Screenshots (if appropriate):

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Rechi commented on docs/README.iOS.md:
`````diff
@@ -143,51 +143,66 @@ make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS
+
+Example to generate Xcode project for ARM 32bit:
+````
+make -C tools/depends/target/cmakebuildsys GEN=Xcode CMAKE_EXTRA_ARGUMENTS="-DCMAKE_TOOLCHAIN_FILE=/Users/Shared/xbmc-depends/iphoneos9.3_armv7-target/share/Toolchain.cmake"
`````
passing `CMAKE_TOOLCHAIN_FILE` is a bad example as this is auto set to the correct path by the cmakebuildsys Makefile

fuzzard replied:
If you look a little closer, the toolchain file in the example is the toolchain file used for the 32bit example in the existing docs. This is how you would use a different toolchain and provides you with a 64bit (default) and info for 32bit, same as existing docs.

Rechi replied:
If you want to build 32 bit, you have to configure depends for 32 bit and it will set the correct CMAKE_TOOLCHAIN_FILE automatically.

fuzzard replied:
so drop that example, and assume they do depends on the appropriate platform in 4?

Rechi replied:
If you don't build depends for the platform you want before building Kodi it won't work anyway.

Here the reason for not setting CMAKE_TOOLCHAIN_FILE explicit.
```
OS X
tools/depends/Makefile.include:16:PREFIX=/Users/Shared/xbmc-depends/macosx10.13_x86_64-target-debug
tools/depends/Makefile.include:63:CMAKE=/Users/Shared/xbmc-depends/x86_64-darwin17.7.0-native/bin/cmake -DCMAKE_TOOLCHAIN_FILE=$(PREFIX)/share/Toolchain.cmake -DCMAKE_INSTALL_PREFIX=$(PREFIX)

iOS 32-Bit
tools/depends/Makefile.include:16:PREFIX=/Users/Shared/xbmc-depends/iphoneos11.3_armv7-target-debug
tools/depends/Makefile.include:63:CMAKE=/Users/Shared/xbmc-depends/x86_64-darwin17.7.0-native/bin/cmake -DCMAKE_TOOLCHAIN_FILE=$(PREFIX)/share/Toolchain.cmake -DCMAKE_INSTALL_PREFIX=$(PREFIX)

iOS 64-Bit
tools/depends/Makefile.include:16:PREFIX=/Users/Shared/xbmc-depends/iphoneos11.3_arm64-target-debug
tools/depends/Makefile.include:63:CMAKE=/Users/Shared/xbmc-depends/x86_64-darwin17.7.0-native/bin/cmake -DCMAKE_TOOLCHAIN_FILE=$(PREFIX)/share/Toolchain.cmake -DCMAKE_INSTALL_PREFIX=$(PREFIX)
```

fuzzard replied:
Fair call on that. Ill remove it. do you want/need an explicit 32bit build, or can the assumption be that if they wanted 32bit, they would have built 32bit depends, and then straight cmakebuildsys call will do the right toolchain?

Rechi replied:
I would keep the explicit cmake configure commands.
For the cmakebuildsys commands there is no need to mention CMAKE_TOOLCHAIN_FILE.

fuzzard replied:
K, have pushed a change. Not sure if its too long winded now as a single section combining xcode/xcodebuild and both cmakebuildsys and old cmake commands

Rechi commented on docs/README.iOS.md:
`````diff
@@ -143,51 +143,66 @@ make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS
+````
 
-**WARNING:** The toolchain file location differs depending on your iOS and SDK version. You have to replace `x86_64-darwin15.6.0-native` and `iphoneos11.3_arm64-target-debug` or `iphoneos9.3_armv7-target` in the paths above with the correct ones on your system.
+**WARNING:** The toolchain file location differs depending on your iOS and SDK version. You have to replace `iphoneos11.3_arm64-target-debug` or `iphoneos9.3_armv7-target` in the paths above with the correct ones on your system.
`````
not needed as it is autodetected

fuzzard replied:
Left it in reference to the 32bit example for the cmake_extra_arguments and the 32bit toolchain file

Rechi replied:
That isn't the correct way for building 32bit

fuzzard replied:
all it does is create the xcodeproj files, same as existing docs. Not sure i understand where the fault is? the produced project is exactly what would be produced by the current docs

Rechi commented on docs/README.iOS.md:
````diff
@@ -143,51 +143,66 @@ make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS
+Generate Xcode project for ARM:
+```
+make -C tools/depends/target/cmakebuildsys GEN=Xcode
+```
````
project was already generated in section 6.1

fuzzard replied:
Is it expected that if someone was doing an xcodebuild they would follow the steps from the xcode build to get to it?

Rechi replied:
yes, otherwise you could remove 6.1, which will lead to the same result (list the configure command once)

fuzzard replied:
Ok, ive had others say they would have thought each section was completely independent of the others. Happy to go your way though. 👍

Rechi replied:
Then how to clone, configure depends and build depends would be missing too in this step.

fuzzard replied:
point sections, majors naturally follow on 1-2-3-4, etc, but 6.1, 6.2, 6.3 arent all done.
But as i said, happy to put it back to how existing docs are.

Rechi replied:
I looked now at the whole step 6 and think 6.2 should actually be a subsection of 6.1.

fuzzard replied:
As a named subsection, or just as a mention (eg. either open xcode or use xcodebuild with the following command) in 6.1 and combine them

Rechi replied:
I think something like, you can use this command to build from terminal instead of opening xcode would be enough.

Rechi commented on docs/README.iOS.md:
```diff
@@ -143,51 +143,66 @@ make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS
+
+Additional cmake arguments can be supplied via the CMAKE_EXTRA_ARGUMENTS command line variable
+
+An example to set -DCMAKE_TOOLCHAIN_FILE for an alternate sdk/ios version:
```
bad example

fuzzard replied:
happy to remove it, but the example provided the 32bit option from existing docs.

Rechi replied:
before it was needed, but cmakebuildsys auto sets it

fuzzard replied:
Ok, will remove 👍

Rechi reviewed (changes requested):
The updated documentation do not list any way to have a 32-bit and 64-bit iOS project from the same source at the same time.

fuzzard commented:
Have added a commit to change jenkins to use xcodebuild.

fuzzard commented:
just having a quick look at the buildsteps again, is there an existing jenkins variable we can pull from for target config (debug or release)?

Memphiz commented:
> just having a quick look at the buildsteps again, is there an existing jenkins variable we can pull from for target config (debug or release)?

See tvOS buildsteps in yab branch - variable „Configuration“ should do it

Memphiz commented:
Jenkins build this please

Rechi commented:
Will complete makefile compatibility be kept for iOS and who will check that if Jenkins is switched to Xcode?

fuzzard commented:
up to you guys really. From what kambala/memphiz say in that other PR, if that is merged, it needs xcodebuild for packaging images via the storyboard method. It cant be done with make, therefore there will be no icons/images using make if that is merged.

Do you want to make another ios jenkins platform for make builds?

Ill let you guys decide, the PR is there if you want to go down the xcodebuild route.
Would still like to get the GEN change in regardless of docs/jenkins. We will need it for tvos, so happy to carry it in dev branches until tvos stuff is ready for master if you dont want it for ios right now. But it can be put in with zero changes to jenkins/docs if it doesnt match the status quo.

Memphiz commented:
@Rechi what do we need Makefile support for - till now it was only used for Jenkins. I don‘t know any devs that want to use the Makefile target for Darwin platforms for development because it makes zero sense. On macos you wanna use Xcode anyways no?

Rechi commented:
I mostly used makefiles, as I only did quick compile tests for iOS and OS X, but that's not important.

The important thing is, if makefile support won't be guaranteed anymore, remove it from the docs and autoset the cmake generator to Xcode for iOS. Otherwise you get the impression it is supported and should work.
Also do not forget to remove the no longer needed explicit setting of some environment variables in cmake files.

Memphiz commented:
Agreed ...

Memphiz commented:
We could add a make short cut to Xcode build though to satisfy the „quick compile test“ scenario though ;)

Rechi commented:
No need for that, I just used what cmakebuildsys generated. `cmake --build <PATH> --target <TARGET>` also works with Xcode.

fuzzard commented:
```ifeq ($(OS),ios)
  CMAKE_BUILD_ARGUMENTS = -G Xcode
else
  CMAKE_BUILD_ARGUMENTS = -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=$(Configuration)
endif```

Prefer that rather than the argument?

Rechi commented:
If iOS will only work with Xcode it's not preferred, then it's required.

fuzzard commented:
Before i push a fixup for that, the error issue on the 64bit ios build.

Is it acceptable to add -Wno-shorten-64-to-32 to silence those warnings for ios?
xcode 10+ turns that warning on by default apparently for 64bit builds, and fixing 800+ warnings through the entire kodi codebase is probably not doable right now.

Rechi commented on docs/README.iOS.md:
```diff
@@ -140,27 +139,38 @@ make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS
 
-### 6.1. Build with Xcode
-Create an out-of-source build directory:
+Change to Kodi's source code directory:
```
do not remove the info, for how to do a out of source build

fuzzard replied:
what is preferred by team kodi. a true out of tree build, or cmakesys doing an "out of tree" build in a dir inside source?

i ask only because is it worthwhile adding the following to cmakebuildsys to allow build dir to be provided by user?
```
ifeq ($(BUILD_DIR),)
  BUILD_DIR=$(CMAKE_SOURCE_DIR)/build
endif
```

Rechi replied:
For both are out of source builds, so everything which doesn't hit the following code.
https://github.com/xbmc/xbmc/blob/40da2b6ef0dbee65ddf90cdb375b1ad5bf5e772b/CMakeLists.txt#L20-L22

Your suggested code wouldn't be limited to buildir inside the source tree, so it's okay for me.
If you want to add the BUILD_DIR change, README.iOS.md, README.macOS.md and README.tvOS.md have to be adjusted. Additionally to the Xcode auto detection for iOS add and option to switch macOS to Xcode generator. 
Would be great if you add the info about the optional BUILD_DIR variable in README.Android.md and README.RaspberryPi.md.

fuzzard replied:
Ok, ill try tackle all that over the weekend.

fuzzard replied:
have added a 3rd commit for build_dir change. Have updated other docs as suggested.
Let me know what you think needs altering.

Rechi commented on tools/buildsteps/ios/configure-xbmc:
```diff
@@ -2,4 +2,4 @@ WORKSPACE=${WORKSPACE:-$( cd $(dirname $0)/../../.. ; pwd -P )}
 . $WORKSPACE/tools/buildsteps/defaultenv
 
-make -C $WORKSPACE/tools/depends/target/cmakebuildsys
+make -C $WORKSPACE/tools/depends/target/cmakebuildsys CMAKE_EXTRA_ARGUMENTS="-DCMAKE_XCODE_ATTRIBUTE_GCC_WARN_64_TO_32_BIT_CONVERSION=NO"
```
no, I've adjusted jenkins to accept the current amount of shorten-64-to-32 warnings

fuzzard replied:
ok, ill remove that. felt weird putting it in, but theres no hope of cleaning that warning up anytime soon, and its a standard xcode 10 warning for 64bit targets now.

Rechi reviewed (commented):
The commits have to be squashed as they do not work independent.

Memphiz commented:
Jenkins build this please

fuzzard commented:
@Rechi As far as i can see, the individual commits can all stand alone, but happy to squash into one.
Ive got one minor fix to do regarding the tvos.md doc (C&P error with ios info).

Do you want me to squash everything into a single commit?

Rechi commented:
@fuzzard please split this into two separate commits
1. BUILD_DIR changes in docs and cmakebuildsys
2. Xcode genrator changes in docs, buildsteps and cmakebuildsys

fuzzard committed:
[cmakebuildsys] Add build_dir option + Docs

Allow supplying BUILD_DIR option to cmakebuildsys to define build location. Default will
fall back to <source>/build if not supplied.

fuzzard committed:
[cmakebuildsys\ios] generator option (GEN) + docs

Xcode is now the only supported cmake generator for ios. make build no longer available
for ios, removed from docs
updated jenkins buildsteps for ios to use xcodebuild
macos/tvos docs for cmakebuildsys updated for use with xcodebuild
Cleanup of cmake and darwin/support scripts to use existing default variables from xcode
project instead of defining additional.

fuzzard commented:
Done as requested.

Please note, ios/tvos/macos build_dir docs are in the 2nd commit as they currently dont have any mention of cmakebuildsys. I think its better to add build_dir info after/with the introduction of cmakebuildsys usage for those 3 platforms.

Rechi commented:
cmakebuildsys is mentioned in iOS, macOS and tvOS README
```
git grep cmakebuildsys upstream/master -- docs/
upstream/master:docs/README.Android.md:make -C tools/depends/target/cmakebuildsys
upstream/master:docs/README.RaspberryPi.md:make -C tools/depends/target/cmakebuildsys
upstream/master:docs/README.iOS.md:make -C tools/depends/target/cmakebuildsys
upstream/master:docs/README.macOS.md:make -C tools/depends/target/cmakebuildsys
upstream/master:docs/README.tvOS.md:make -C tools/depends/target/cmakebuildsys
```

fuzzard commented:
Ok, ios removes the make build feature as part of second commit. Do you want the builddir added and then moved, or can the single doc change be done in the second commit?
tvos also cannot be built with make at all. Honestly half of that doc shouldnt be in there as its misleading currently (binary-addons, actual kodi build, packaging, etc isnt in master).

I can definitely do the macos change, ill fix that now

Rechi commented on docs/README.tvOS.md:
```diff
@@ -132,67 +134,58 @@ make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS
+
+**NOTE:** The below is not currently possible in Kodi master repo
+
+Before you can use Xcode to build Kodi, the Xcode project has to be generated with CMake. CMake is built as part of the dependencies and doesn't have to be installed separately. A toolchain file is also generated and is used to configure CMake.
```
This is already in the readme 4 lines below.

Rechi commented on docs/README.tvOS.md:
````diff
@@ -132,67 +134,58 @@ make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS
 ```
 
-**WARNING:** The toolchain file location differs depending on your tvOS and SDK version. You have to replace `x86_64-darwin18.5.0-native` and `appletvos12.2_arm64-target-debug` in the paths above with the correct ones on your system.
+**WARNING:** The toolchain file location differs depending on your iOS and SDK version. You have to replace `x86_64-darwin18.5.0-native` and `appletvos12.2_arm64-target-debug` in the paths above with the correct ones on your system.
````
this is the tvOS readme

Rechi commented:
wrong revision was build
jenkins build this please

Razzeee commented:
jenkins build this please

lrusak commented:
jenkins build this please

fuzzard commented:
need a little cmake help here i think. with this ios 32bit build error

```
15:24:40 CMake Error at cmake/modules/FindFmt.cmake:64 (find_package):
15:24:40   Could not find a configuration file for package "FMT" that is compatible
15:24:40   with requested version "3.0.1".
15:24:40 
15:24:40   The following configuration files were considered but not accepted:
15:24:40 
15:24:40     /Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends/iphoneos12.2_armv7-target-debug/lib/cmake/fmt/fmt-config.cmake, version: 5.1.0 (32bit)
15:24:40 
15:24:40 Call Stack (most recent call first):
15:24:40   cmake/scripts/common/Macros.cmake:366 (find_package)
15:24:40   cmake/scripts/common/Macros.cmake:380 (find_package_with_ver)
15:24:40   CMakeLists.txt:171 (core_require_dep)
15:24:40 
15:24:40 
15:24:40 -- Configuring incomplete, errors occurred!
```

So locally, i was able to change cmake/modules/FindFmt.cmake to the following as a test

```
if(CORE_SYSTEM_NAME STREQUAL windows OR CORE_SYSTEM_NAME STREQUAL windowsstore)
  # TODO: fix windows fmt package to include fmt-config.cmake and fmt-config-version.cmake
  set(FMT_VERSION 3.0.1)
elseif(CORE_SYSTEM_NAME STREQUAL ios)
  set(FMT_VERSION 5.1.0)
else()
  find_package(FMT 3.0.1 CONFIG REQUIRED QUIET)
endif()
```

Note i specifically added the 5.1.0 version number as it is the current version of libfmt dependency (https://github.com/xbmc/xbmc/commit/f3924c3447f67bb4d90a6332fbaa3ae50c47b717)
This enabled the correct libfmt to be found. 
libfmt is built in target during depends build correctly, and is installed to /Users/Shared/xbmc-depends/iphoneos12.2_armv7-target-debug/lib/ correctly. The .a file reports the correct architecture with lipo

```
$ lipo -info libfmt.a 
Non-fat file: libfmt.a is architecture: armv7
```

/Users/Shared/xbmc-depends/iphoneos12.2_armv7-target-debug/lib/cmake/fmt has the following cmake files after the install step of the dependency build

```
-rw-r--r--  1   1386 10 Jul 11:22 fmt-config-version.cmake
-rw-r--r--  1    962 10 Jul 11:22 fmt-config.cmake
-rw-r--r--  1    795 10 Jul 11:22 fmt-targets-release.cmake
-rw-r--r--  1   3646 10 Jul 11:22 fmt-targets.cmake
```


Anyone able to give advice on a proper and correct fix to get find_package to correctly locate and use libfmt?

Rechi commented on tools/depends/configure.ac:
```diff
@@ -367,7 +367,7 @@ case $host in
 
         if [ ! test "x$use_cpu" = "xarm64" ]; then
-          platform_cflags+=" -mcpu=cortex-a8 -mfpu=neon"
+          platform_cflags+=" -mfpu=neon"
```
removing `-mcpu=cortex-a8` is wrong, it is supported for armv7 builds

fuzzard replied:
xcode 10 errors with unknown mcpu.

Ill get a local build going to get the exact error, but if you can locate the previous jenkins build, it will have the error

fuzzard replied:
Found the log - https://jenkins.kodi.tv/job/IOS/10085/console

```
12:22:25     clang: error: the clang compiler does not support '-mcpu=cortex-a8'
12:22:25     
12:22:25     ** BUILD FAILED **
12:22:25     
12:22:25     
12:22:25     The following build commands failed:
12:22:25     	CompileC CMAKE_TRY_COMPILE.build/Debug-iphoneos/cmTC_32c45.build/Objects-normal/arm64/testCCompiler.o testCCompiler.c normal arm64 c com.apple.compilers.llvm.clang.1_0.compiler
12:22:25     (1 failure)
12:22:25     
12:22:25 
12:22:25   
12:22:25 
12:22:25   CMake will not be able to correctly generate this project.
12:22:25 Call Stack (most recent call first):
12:22:25   CMakeLists.txt:2 (project)
12:22:25 
12:22:25 
12:22:25 -- Configuring incomplete, errors occurred!
```

Rechi replied:
You cut the important thing, the invoked command.

`/Applications/Xcode10.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c -arch arm64 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-float-conversion -Wno-non-literal-null-conversion -Wno-objc-literal-conversion -Wshorten-64-to-32 -Wpointer-sign -Wno-newline-eof -DCMAKE_INTDIR=\"Debug-iphoneos\" -isysroot /Applications/Xcode10.2.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk -fstrict-aliasing -Wdeprecated-declarations -miphoneos-version-min=12.2 -g -Wno-sign-conversion -Wno-infinite-recursion -Wno-comma -Wno-block-capture-autoreleasing -Wno-strict-prototypes -Wno-semicolon-before-method-body -I/Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/Debug/include -I/Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/CMAKE_TRY_COMPILE.build/Debug-iphoneos/cmTC_32c45.build/DerivedSources-normal/arm64 -I/Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/CMAKE_TRY_COMPILE.build/Debug-iphoneos/cmTC_32c45.build/DerivedSources/arm64 -I/Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/CMAKE_TRY_COMPILE.build/Debug-iphoneos/cmTC_32c45.build/DerivedSources -Wmost -Wno-four-char-constants -Wno-unknown-pragmas -F/Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/Debug -fheinous-gnu-extensions -no-cpp-precomp -mcpu=cortex-a8 -mfpu=neon -ftree-vectorize -pipe -Wno-trigraphs -fpascal-strings -Wreturn-type -Wunused-variable -fmessage-length=0 -Wno-error=implicit-function-declaration -arch armv7 -miphoneos-version-min=9.0 -D_DEBUG -isysroot /Applications/Xcode10.2.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk -isystem /Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends/iphoneos12.2_armv7-target-debug/include -fheinous-gnu-extensions -no-cpp-precomp -mcpu=cortex-a8 -mfpu=neon -ftree-vectorize -pipe -Wno-trigraphs -fpascal-strings -Wreturn-type -Wunused-variable -fmessage-length=0 -Wno-error=implicit-function-declaration -arch armv7 -miphoneos-version-min=9.0 -D_DEBUG -isysroot /Applications/Xcode10.2.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk -isystem /Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends/iphoneos12.2_armv7-target-debug/include -MMD -MT dependencies -MF /Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/CMAKE_TRY_COMPILE.build/Debug-iphoneos/cmTC_32c45.build/Objects-normal/arm64/testCCompiler.d --serialize-diagnostics /Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/CMAKE_TRY_COMPILE.build/Debug-iphoneos/cmTC_32c45.build/Objects-normal/arm64/testCCompiler.dia -c /Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/testCCompiler.c -o /Users/Shared/jenkins/workspace/IOS/build/CMakeFiles/CMakeTmp/CMAKE_TRY_COMPILE.build/Debug-iphoneos/cmTC_32c45.build/Objects-normal/arm64/testCCompiler.o`

You see the wrong argument?

fuzzard replied:
ive no idea how/where it switches to arm64 for that test though. The below is what the configure command sets in that build failure.

```
11:06:24 #------- configuration -------#
11:06:24 -e ccache:		 yes
11:06:24 -e build type:	 debug
11:06:24 -e toolchain:	 /Applications/Xcode10.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain
11:06:24 -e cpu:		 armv7
11:06:24 -e host:		 arm-apple-darwin
11:06:24 -e cflags:		 -fheinous-gnu-extensions -no-cpp-precomp -mcpu=cortex-a8 -mfpu=neon -ftree-vectorize -pipe -Wno-trigraphs -fpascal-strings -Wreturn-type -Wunused-variable -fmessage-length=0 -gdwarf-2 -Wno-error=implicit-function-declaration -arch armv7 -miphoneos-version-min=9.0 -g -D_DEBUG 
11:06:24 -e cxxflags:	 -std=c++11 -no-cpp-precomp -arch armv7 -miphoneos-version-min=9.0 -stdlib=libc++ -g -D_DEBUG 
11:06:24 -e ldflags:	 -Wl,-search_paths_first -Wl,-segalign,4000 -L/Applications/Xcode10.2.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk/usr/lib/system -arch armv7 -miphoneos-version-min=9.0 -isysroot /Applications/Xcode10.2.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk -stdlib=libc++  -liconv 
11:06:24 -e ffmpeg options:	 
11:06:24 -e prefix:		 /Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends
11:06:24 -e depends:	 /Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends/iphoneos12.2_armv7-target-debug
```

fuzzard replied:
cmakeoutput.log

https://paste.kodi.tv/ukecevakuj

cmakeerror.log

https://paste.kodi.tv/azaseyajax

Rechi replied:
I invoked the command to generated the Xcode project for armv7 and after it failed searched for arm64 in the build folder.
```
$ grep --color -R arm64 *
CMakeFiles/3.12.2/CMakeCXXCompiler.cmake:set(CMAKE_XCODE_ARCHS "arm64")
[...]
```

fuzzard replied:
yeah, looking at that right now. going down the path of a cmake issue? i really dont know at this point sorry. doing some searching now.

fuzzard replied:
can add the following to toolchain.cmake, and it forces it to skip compiler tests. i know thats a heavy handed way around it, but just documenting another method for now.

```
set(CMAKE_C_COMPILER_FORCED TRUE)
```

no need to force CMAKE_CXX_COMPILER, as it either doest test, or doesnt fail the test.

fuzzard replied:
even with that (keeping the -mcpu=cortex-a8 flag), libfmt still fails with the same information provided above.

Rechi replied:
That's because `set(CMAKE_C_COMPILER_FORCED TRUE)` is a hack. You have to tell cmake to configure for armv7. Do not try to tell cmake to skip compiler checks.

fuzzard replied:
as far as i can see, all the things i can find are saying its already been told to build for armv7. Ive no idea why that test does what it does.

Kambala has confirmed both the cmake test hack and the cortex-a8 removal and the libfmt fix end in runnable builds on ios9 devices.

Honestly, i see 32bit ios as a niche. I know you might not see it that way, but 32bit in the apple world is/has gone the way of the dodo. unless someone can point me in the right direction for an acceptable solution, i personally dont think the time invested so far for it is worth it.

Rechi replied:
No, otherwise the compiler check wouldn't pass `-arch arm64` as argument.

I never said it doesn't work, only that the this is a hack and therefore not acceptable.

If you think that's the only way, okay. In that case you have to do a complete removal, only stop building for ios armv7 it isn't enough.

Memphiz replied:
So I tend to drop armv7 support for iOS then.

fuzzard replied:
No longer needed, thanks for the fix Rechi

fuzzard commented:
@Rechi Have discussed with a couple of people, and ive started down removing 32bit ios in a local branch.
What im asking here is if you want that in a separate PR, or bundled into this with an edit to the PR message?

fuzzard commented:
Have added an ios32bit removal commit.
Please note, ive bumped the iphoneos-version-min to 11.0 from 9.0. This is really an arbitrary change. Im not totally sure about the 64bit support on devices that support ios9, but i know 10+ all run 64bit apps fine. May want to drop it down to 10, but either is fine.

Memphiz commented:
I‘d go for 10 then ...

Rechi commented:
The minimum iOS version has to be adjusted at the following places too.

https://github.com/xbmc/xbmc/blob/5d7938dcba9700c9d41eeea24b88e2f7201e089d/cmake/scripts/ios/ArchSetup.cmake#L44
https://github.com/xbmc/xbmc/blob/57557199099ea7e511e8ff0073505b912f2d6cb7/tools/darwin/packaging/ios/mkdeb-ios.sh.in#L87
https://github.com/xbmc/xbmc/blob/9fcaef0ed9c4db8ca3775578a1e909ce21aee96d/xbmc/platform/darwin/ios/Info.plist.in#L72-L73

fuzzard commented:
I saw your other PR. Does that get iOS 32bit building? If so, do you want that drop commit removed?

Rechi commented:
Yes, that PR fixes generating the Xcode project for iOS armv7.
If you still want to drop iOS arm support it has to be done in a seperate PR, as it's no longer required for this PR.

kambala-decapitator commented:
we'll need to make a cleanup of `Info.plist.in` in this PR, as Xcode inserts some keys on its own.

fuzzard commented:
more than happy to keep ios32bit @Rechi 
Have dropped that from this commit.

Regarding Kambala's comment, he has sent through a copy of what he wants changed, but i wasnt sure if it was absolutely required in this PR, or if it should just be done in a followup PR. there are a few areas in that plist include that could use some updating.

From my build testing, everything builds and runs ok without the current plist, but open to whatever is required. 

If someone could kick off a jenkins build, hopefully this will give all greens after Rechi's PR for armv7 was merged.

notspiff commented:
jenkins build this please

Rechi commented on tools/depends/target/cmakebuildsys/Makefile:
```diff
@@ -10,11 +10,21 @@ ifeq ($(Configuration),)
 endif
 
-BUILD_DIR=$(CMAKE_SOURCE_DIR)/build
+ifeq ($(OS),ios)
```
`ifeq ($(OS),darwin_embedded)`

fuzzard replied:
Done

fuzzard commented:
Rebased, if someone wouldnt mind a jenkins kick

phunkyfish commented:
Can anyone give this build a kick?

DaveTBlake commented:
jenkins build this please

phunkyfish commented:
There may be a Jenkins problem here. That’s doesn’t appear to have worked at all.

phunkyfish commented:
Windows appears to be failing in general for kodi and pvr builds.

phunkyfish commented:
Woohoo, we have a green build ;)

Is this ready to go in?

Rechi commented:
Cmake code which is obsolete, becauase Xcode is the only supported generator for iOS has to be removed too in this PR.

phunkyfish commented:
Ok, @fuzzard back to you.

fuzzard commented:
@Rechi i could only find the one ios reference regarding make in the cmake files. 
The only other one is in the macros.cmake located in cmake/scripts/osx, but is a symlink used for ios. This means cant drop the reference to make in there as osx still uses make.

Rechi commented:
You are talking about the following line? Are you sure it can't be adjusted?
https://github.com/xbmc/xbmc/blob/3603621e010516de043e0811431ec37d59bc7104/cmake/scripts/osx/Macros.cmake#L23

Also some of the following explicit set variables can be removed, as they are set by Xcode anyway.
https://github.com/xbmc/xbmc/blob/3603621e010516de043e0811431ec37d59bc7104/cmake/scripts/darwin_embedded/Install.cmake#L61-L86

fuzzard commented:
Cheers, ill look into it

fuzzard commented:
Please note, there is now a new commit added to this.
With the removal of the make build for ios, there wasnt a way (that i have found) that can build, and package, binary addons.
This adds binary-addons as a dependency of the kodi target, and defaults to build all binary addons. I have included documentation on how to selectively choose addons to build to match the prior binary-addon creation during the xcode generator.

I have not altered the jenkins build steps for binary addons, as they will currently build with its existing steps. May wish to alter it to use the commited method, but for now i thought it would be best for feedback before i blindly try to do that.

Any feedback on the cmake changes would be greatly appreciated. Cmake isnt my cup of tea, so i hope this method is appropriate.

Havent looked into the last comment yet Rechi, sorry. Still on my todo list for this PR.

phunkyfish commented:
From the docs I don’t see how a user would skip building the binary addons.

Rechi commented:
`make -C tools/depends/target/binary-addons` is the command to build all add-ons, see build docs.

fuzzard commented:
Binary add-ons get built into Kodi/addons using that, yes, but for non make file generation(ie Xcode) they don't get copied into the build folder, so are never packaged when the Xcode project is built. I couldn't find where/why that occurs, so hence this implementation.
As I have said, cmake isn't my thing, so any guidance on why the above occurs, happy to change/revert the commit.

Rechi commented:
I already told why it works with Jenkins (https://forum.kodi.tv/showthread.php?tid=343197&pid=2847828#pid2847828) and it's not related to any CMake generator.

In the current way this is not acceptable.

fuzzard commented:
Ok, that's new info to me, fine. What about just adding the build_dir change in cmakebuildsys to binary_addons to set install_prefix to default $kodi/build if not supplied if darwin_embedded target? 
I assume some targets still want/need into $kodi/add-ons for some reason?

phunkyfish commented:
I think just add INSTALL_DIR to the binary addons build command should do it. It just mean the generated directory needs to exist before building the addons.

fuzzard commented:
thats one way, but my thinking is sane defaults should provide what is expected from a default build. Yes you can have that as a argument in the build doc that gets the addons into the correct place, but why not just have make -C tools/depends/target/binary-addons put in the correct place by default. And all of that can be encased by a check for the platforms it affects (darwin_embedded). Just simplifies the steps required?

> In the current state this is not acceptable.

Im understanding of this, but out of curiosity, is it the concept of an all encompassed single xcode build project that isnt acceptable, or just the piss poor implementation?
I just imagine that after building depends, a dev/user then runs one command (cmakebuildsys), and then opens xcode and presses build. It gives all the default settings(addons built, packaged, ready for run). Just seems nice and clean to me.

fuzzard commented:
No, not at all. All that behaviour is still there, nothing has changed at all. The only thing that gets changed is you don't have to build add-ons prior, it can/will be built in the Kodi project build.
If you read the docs I added, it still gives the builder the option to do all,some,none with the same flexibility that is currently there.

phunkyfish commented:
Ya, but you have to build kodi. What if you only need to build an addon or a set of them?

fuzzard commented:
no you don't, it adds a target in the Xcode project, select binary-addons -> generic iOS device, and it only builds the add-ons made when generating the project.

phunkyfish commented:
Then I’m out of reasons 😉

fuzzard commented:
Any better @Rechi ?

fuzzard commented:
Have updated install.cmake and the support shell scripts to use xcode project defined variables.

phunkyfish commented:
@fuzzard is the only thing outstanding how binary addons are built?

fuzzard commented:
Yeah, just need some feedback if that Xcode commit can go in or not I guess.

Rechi commented:
Please drop the binary-addons commit. It should be discussed in a separate PR.

fuzzard commented:
no worries, done.

Rechi commented:
Please also squash the cmakebuildsys generator and xcode variables commits.

fuzzard commented:
all done

Rechi commented:
wrong revision was build
jenkins build this please

fuzzard commented:
jenkins build this please

phunkyfish commented:
Ok, is this ready to go in now.

phunkyfish commented:
@Rechi are you happy with this now?

phunkyfish commented:
Can this go in? I don't believe there is anything outstanding and it's blocking for a lot of the tvOS work so would be great to have it in there.

phunkyfish commented:
@Rechi are happy with this now? Commit messages updated.
